src/main/java/com/odysseusinc/arachne/portal/service/analysis/impl/AnalysisFilesSavingServiceImpl.java

```diff
@@ -244,38 +251,31 @@ public AnalysisFile saveFileByLink(String link, IUser user, A analysis, String l
-        StringBuilder generatedFileNameBuilder = new StringBuilder();
-        generatedFileNameBuilder.append(CommonAnalysisType.COHORT.getTitle());
-        try (final ZipOutputStream zos = new ZipOutputStream(out)) {
-            files.stream()
+    public void saveCohortAnalysisArchive(A analysis, DataReference dataReference, IUser user, List<MultipartFile> files) {
+
+        try (ByteArrayOutputStream out = new ByteArrayOutputStream();
+             ZipOutputStream zos = new ZipOutputStream(out)) {
+            MultipartFile genericSqlFile = files.stream()
+                    .filter(file -> file.getName().endsWith(OHDSI_SQL_EXT))
+                    .findAny()
+                    .orElseThrow(() -> new ArachneSystemRuntimeException(String.format("There is no sql file for %s analysis.", analysis.getId())));
+            generateFilesForEachDialectAndAddToZip(zos, genericSqlFile);
+
+            Collection<MultipartFile> allFilesWithoutSqlFile = files.stream()
+                    .filter(file -> ObjectUtils.notEqual(file, genericSqlFile))
                     .filter(file -> !ANALYSIS_INFO_FILE_DESCRIPTION.equals(file.getName()))
-                    .forEach(file -> {
-                        try {
-                            if (file.getName().endsWith(OHDSI_SQL_EXT)) {
-                                String shortBaseName = generateDialectVersions(zos, file);
-                                generatedFileNameBuilder.append("_");
-                                generatedFileNameBuilder.append(shortBaseName);
-                            } else {
-                                String fileName = file.getName();
-                                ZipUtil.addZipEntry(zos, fileName, file.getInputStream());
-                            }
-                        } catch (IOException e) {
-                            throw new ArachneSystemRuntimeException("Failed to add file to archive: " + file.getName(), e);
-                        }
-                    });
-        } catch (IOException e) {
-            throw new ArachneSystemRuntimeException(e);
-        }
-        String fileName = generatedFileNameBuilder.append(".zip").toString();
-        final MultipartFile sqlArchive = new MockMultipartFile(fileName, fileName, "application/zip",
-                out.toByteArray());
-        try {
+                    .collect(Collectors.toList());
+            ZipUtil.addZipEntries(zos, allFilesWithoutSqlFile);
+
+            String analysisName = getAnalysisName(genericSqlFile);
+            String fileName = AnalysisArchiveUtils.getArchiveFileName(CommonAnalysisType.COHORT, analysisName);
+            MultipartFile sqlArchive = new MockMultipartFile(fileName, fileName, com.google.common.net.MediaType.ZIP.toString(),
+                    out.toByteArray());
             saveFile(sqlArchive, user, analysis, fileName, false, dataReference);
-        } catch (AlreadyExistException e) {
-            log.error("Failed to save file", e);
+
```

Comment:
There are some behavior changes during refactoring. Now it tries to save output stream before zip stream actually closed. Looks suspicious and might be the reason of some unexpected errors.

Reply:
hmm, this one is good, I would think about this

Reply:
That is interesting. According theByteArrayOutputStream jdoc, the close method does not affect anything: 

From ByteArrayOutputStream class:
```    
/**
     * Closing a <tt>ByteArrayOutputStream</tt> has no effect. The methods in
     * this class can be called after the stream has been closed without
     * generating an <tt>IOException</tt>.
     */
    public void close() throws IOException {
    }

```

Reply:
Well, but my concern is more about `ZipOutputStream`

Reply:
fixed